pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -229,6 +223,21 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun)
 		logger.Errorf("Reconcile error: %v", err.Error())
 	}
 
+	if pr.IsDone() {
+		// We get latest pipelinerun cr already to avoid recount
+		newPr, err := c.pipelineRunLister.PipelineRuns(pr.Namespace).Get(pr.Name)
```

Comment:
I'm not super familiar with how/when the lister updates, but wouldn't pulling in the latest status mean that the current pr status and the "before" status always match, since the status is updated during `reconcile`?

It's would be useful to add a test to make sure we're handling the reconciler state changes properly (i.e. making sure we're avoiding instances where before always equals after) - right now we're really only testing the DurationAndCount func directly.

Reply:
`reconcile` changes status but that isn't applied to the `pipelienrun` object in  `Informer` cache.

Reply:
My understanding aligns with @khrm, the status changes but the change is not visible through the lister.

Reply:
For my own knowledge - When would the lister value update? Is it because we're updating only the status that this is safe to do?

I'm a bit confused with the current impl/comments since we're getting the "latest" pipelinerun, but this is really the pipelinerun before the latest status update?

I played around with this change and this appears to be correct (I even tried adding some sleeps to try and trigger a race condition), but it feels like we're relying on non-obvious behavior of the lister. I'd have a lot more peace of mind if we could add a test to ensure that it's safe rely on this lister behavior and be able to catch if this assumption is no longer true (i.e. if we make an unrelated change that would cause the lister to start returning the true latest value).

Reply:
sure, sounds good. @khrm please add a unit test including an update during the `reconcile` in addition to what we have otherwise looks good to me 👍  Thanks!

Reply:
Please check this comment.

https://github.com/tektoncd/pipeline/pull/4468#issuecomment-1108299026